agent/engine/docker_task_engine_test.go

```diff
@@ -1722,6 +1722,81 @@ func TestPullAndUpdateContainerReference(t *testing.T) {
 	assert.Equal(t, dockerapi.DockerContainerMetadata{}, metadata, "expected empty metadata")
 }
 
+// TestPullAndUpdateContainerReferenceWithCachedImage checks whether a container is added to task engine state when
+// the container is an essential container, the image is cached and DependentContainersPullUpfront is enabled.
+func TestPullAndUpdateContainerReferenceWithCachedImage(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.TODO())
+	defer cancel()
+	cfg := &config.Config{
+		DependentContainersPullUpfront: config.BooleanDefaultFalse{Value: config.ExplicitlyEnabled},
+	}
+	ctrl, client, _, privateTaskEngine, _, imageManager, _ := mocks(t, ctx, cfg)
+	defer ctrl.Finish()
+
+	taskEngine, _ := privateTaskEngine.(*DockerTaskEngine)
+	taskEngine._time = nil
+	imageName := "image"
+	taskArn := "taskArn"
+	container := &apicontainer.Container{
+		Type:      apicontainer.ContainerNormal,
+		Image:     imageName,
+		Essential: true,
+	}
+	task := &apitask.Task{
+		Arn:        taskArn,
+		Containers: []*apicontainer.Container{container},
+	}
+
+	client.EXPECT().PullImage(gomock.Any(), imageName, nil,
+		gomock.Any()).Return(dockerapi.DockerContainerMetadata{Error: dockerapi.CannotPullContainerError{fmt.Errorf("error")}})
+	client.EXPECT().InspectImage(imageName).Return(nil, nil)
+	imageManager.EXPECT().RecordContainerReference(container)
+	imageManager.EXPECT().GetImageStateFromImageName(imageName).Return(nil, false)
+	metadata := taskEngine.pullAndUpdateContainerReference(task, container)
+	pulledContainersMap, ok := taskEngine.State().PulledContainerMapByArn(taskArn)
+	require.True(t, ok, "no container found in the agent state")
+	require.Len(t, pulledContainersMap, 1)
+	assert.Equal(t, dockerapi.DockerContainerMetadata{Error: dockerapi.CannotPullContainerError{fmt.Errorf("error")}},
+		metadata, "expected metadata with error")
+}
+
+// TestPullAndUpdateContainerReferenceFailedToGetImage checks whether a container is added to task engine state when
+// the container is an essential container, ImagePullBehavior is set to always and DependentContainersPullUpfront is enabled.
+func TestPullAndUpdateContainerReferenceFailedToGetImage(t *testing.T) {
```

Comment:
I'd probably use a test table as the code seems largely duplicated.

Reply:
Updated to the table driven test [here](https://github.com/aws/amazon-ecs-agent/pull/2800/files#diff-aa8aba86b2dbb7c182fa55d44aed378f40e350d5ae11405e017dd56d6b58ad04R1689-R1706). Thank you!